### PR TITLE
Do not normalize data when emitting to socket.io sockets

### PR DESCRIPTION
**Pull request by Frank3K**
Thanks for keeping this package maintained!

This PR is a follow-up from https://github.com/thoov/mock-socket/pull/246, where this issue has already been discussed 1.5 years ago. We lost track of that PR (by @thomasdunn) and want to submit it again in a proper form such that it can merged in master.

Note that this code has been running in our own fork for about 1.5 years without any issues.

### CODE REVIEW

File: src/server.js [hunk 1]
```diff
@@ -139,35 +140,37 @@ class Server extends EventTarget {
...
-    } else {
-      data = normalizeSendData(data);
-    }
-
-    websockets.forEach(socket => {
-      if (Array.isArray(data)) {
+    const nonNormalizedData = typeof options !== 'object' || arguments.length > 3
+      ? Array.prototype.slice.call(arguments, 1, arguments.length)
+      : data;
+    const normalizedData = typeof options !== 'object' || arguments.length > 3
```
Comment (Atrue) [inline src/server.js:146]: you have two identical ternary conditions and slice extraction. You can create another variable for the condition and remove the duplicate Array calls. Or maybe you can use the usual if/else construction:
```
let normalizedData;

if (typeof options !== 'object' || arguments.length > 3) {
  data = Array.prototype.slice.call(arguments, 1, arguments.length);
  normalizedData = data.map(item => normalizeSendData(item));
} else {
  normalizedData = normalizeSendData(data);
}
```

File: src/server.js [hunk 2]
```diff
@@ -139,35 +140,37 @@ class Server extends EventTarget {
...
             type: event,
-            data,
+            data: _data,
             origin: this.url,
             target: socket.target
           })
         );
       }
-    });
+    })(socket instanceof SocketIO ? nonNormalizedData : normalizedData));
```
Comment (Atrue) [inline src/server.js:173]: Self-Invoking Functions look weird and hard to read. Please create another variable instead before passing to the dispatchEvent

File: src/socket-io.js [hunk 3]
```diff
@@ -11,7 +11,7 @@ import { createEvent, createMessageEvent, createCloseEvent } from './event/facto
  *
  * http://socket.io/docs/
  */
-class SocketIO extends EventTarget {
+export class SocketIO extends EventTarget {
```
Comment (Atrue) [inline src/socket-io.js:14]: this class is already exported as default

**Review by Atrue — CHANGES_REQUESTED** [review]

File: src/socket-io.js (hunk 3 above)
Comment (Frank3K) [reply src/socket-io.js:14]: The default export is the `IO` function, which returns a new SocketIO instance:

https://github.com/thoov/mock-socket/blob/f1d5de4e6409742670303b6666d833cb4f3af033/src/socket-io.js#L280-L282

The exported `SocketIO` class is used in an `instanceof` check. We can't use the `IO` export for that, right?

File: src/server.js (hunk 2 above)
Comment (Frank3K) [reply src/server.js:173]: I agree. This was based on https://github.com/thoov/mock-socket/pull/246/files#r253630182, but I'll change it to something that's easier to read.

File: src/socket-io.js (hunk 3 above)
Comment (Atrue) [reply src/socket-io.js:14]: Ah, I see. So it's ok, thanks

File: src/server.js [hunk 4]
```diff
@@ -139,29 +140,31 @@ class Server extends EventTarget {
...
-      data = data.map(item => normalizeSendData(item));
+      normalizedData = data.map(item => normalizeSendData(item));
     } else {
-      data = normalizeSendData(data);
+      normalizedData = normalizeSendData(data);
     }
 
     websockets.forEach(socket => {
-      if (Array.isArray(data)) {
+      const messageData = socket instanceof SocketIO ? data : normalizedData;
```
Comment (Frank3K) [inline src/server.js:152]: Before, a `_data` parameter was used. In this new setup `const _data` would issue a [no-underscore-dangle](https://eslint.org/docs/latest/rules/no-underscore-dangle) warning.

**Review by Atrue — APPROVED** [review]

**Comment by Frank3K** [comment]
Thanks for your quick response. Appreciated.

**Comment by Atrue** [comment]
Released in `9.2.1`

**Pull request merged**